### ISSUE TITLE
No way to do ProjectionPlot on non-SPH demeshed particle data

### ISSUE THREAD

**Issue by sheagk** [yt-project/yt#2072]
### Bug report

**Bug summary**

Asking for a projection plot on non-SPH particle data in yt-4.0 (i.e. demeshed yt) suggests using a ParticlePlot (which does not provide the same visual or scientific appeal), or fails if you try to push it too hard.

**Code for reproduction**

Once loading particle data as `ds`, here accomplished with `yt.load_particles`:

```python

prj = yt.ProjectionPlot(ds, 0, ('star', 'particle_mass'), center=center)
   ...:
---------------------------------------------------------------------------
YTInvalidFieldType                        Traceback (most recent call last)
<ipython-input-8-f0cebe42ef5b> in <module>
----> 1 prj = yt.ProjectionPlot(ds, 0, ('star', 'particle_mass'), center=center)

~/yt4/src/yt/yt/visualization/plot_window.py in __init__(self, ds, axis, fields, center, width, axes_unit, weight_field, max_level, origin, right_handed, fontsize, field_parameters, data_source, method, proj_style, window_size, aspect)
   1462         # plotting classes to avoid an exception
   1463         test_data_source = ds.all_data()
-> 1464         validate_mesh_fields(test_data_source, fields)
   1465
   1466         if isinstance(ds, YTSpatialPlotDataset):

~/yt4/src/yt/yt/visualization/plot_window.py in validate_mesh_fields(data_source, fields)
    147
    148     if len(invalid_fields) > 0:
--> 149         raise YTInvalidFieldType(invalid_fields)
    150
    151

YTInvalidFieldType:
SlicePlot, ProjectionPlot, and OffAxisProjectionPlot can only plot fields that
are defined on a mesh or for SPH particles, but received the following N-body
particle fields:

    [('star', 'particle_mass')]

Did you mean to use ParticlePlot or plot a deposited particle field instead?
```

or, alternatively (and probably what _should_ work in the future):

```python

prj = yt.ProjectionPlot(ds, 0, ('deposit', 'star_density'), center=center)
   ...:
---------------------------------------------------------------------------
YTNonIndexedDataContainer                 Traceback (most recent call last)
<ipython-input-9-81448988a4b4> in <module>
----> 1 prj = yt.ProjectionPlot(ds, 0, ('deposit', 'star_density'), center=center)

~/yt4/src/yt/yt/visualization/plot_window.py in __init__(self, ds, axis, fields, center, width, axes_unit, weight_field, max_level, origin, right_handed, fontsize, field_parameters, data_source, method, proj_style, window_size, aspect)
   1478                            center=center, data_source=data_source,
   1479                            field_parameters=field_parameters, method=method,
-> 1480                            max_level=max_level)
   1481         PWViewerMPL.__init__(self, proj, bounds, fields=fields, origin=origin,
   1482                              right_handed=right_handed, fontsize=fontsize, window_size=window_size,

~/yt4/src/yt/yt/data_objects/construction_data_containers.py in __init__(self, field, axis, weight_field, center, ds, data_source, style, method, field_parameters, max_level)
    446
    447         if not self.deserialize(field):
--> 448             self.get_data(field)
    449             self.serialize()
    450

~/yt4/src/yt/yt/data_objects/construction_data_containers.py in get_data(self, fields)
    248                     self._initialize_projected_units(fields, chunk)
    249                     _units_initialized = True
--> 250                 self._handle_chunk(chunk, fields, tree)
    251         # if there's less than nprocs chunks, units won't be initialized
    252         # on all processors, so sync with _projected_units on rank 0

~/yt4/src/yt/yt/data_objects/construction_data_containers.py in _handle_chunk(self, chunk, fields, tree)
    503     def _handle_chunk(self, chunk, fields, tree):
    504         mylog.debug("Adding chunk (%s) to tree (%0.3e GB RAM)",
--> 505                     chunk.ires.size, get_memory_usage()/1024.)
    506         if self.method == "mip" or self._sum_only:
    507             dl = self.ds.quan(1.0, "")

~/yt4/src/yt/yt/data_objects/data_containers.py in ires(self)
   1575         if self._current_chunk is None:
   1576             self.index._identify_base_chunk(self)
-> 1577         return self._current_chunk.ires
   1578
   1579     @property

~/yt4/src/yt/yt/geometry/geometry_handler.py in cached_func(self)
    267             return getattr(self, n)
    268         if self.data_size is None:
--> 269             tr = self._accumulate_values(n[1:])
    270         else:
    271             tr = func(self)

~/yt4/src/yt/yt/geometry/geometry_handler.py in _accumulate_values(self, method)
    295         for obj in self._fast_index or self.objs:
    296             f = getattr(obj, mname)
--> 297             arrs.append(f(self.dobj))
    298         if method == "dtcoords":
    299             arrs = [arr[0] for arr in arrs]

~/yt4/src/yt/yt/data_objects/particle_container.py in _func_non_indexed(self, *args, **kwargs)
     28 def _non_indexed(name):
     29     def _func_non_indexed(self, *args, **kwargs):
---> 30         raise YTNonIndexedDataContainer(self)
     31     return _func_non_indexed
     32

YTNonIndexedDataContainer: The data container (<class 'yt.data_objects.particle_container.ParticleContainer'>) is an unindexed type.  Operations such as ires, icoords, fcoords and fwidth will not work on it.

```

**Expected outcome**

Should be able to visualize at least the density fields of non-SPH particle data.

**Version Information**

  * Operating System:
  * Python Version:
  * yt version:
  * Other Libraries (if applicable): 

Using yt-4.0dev0, merge-rev 8154cbffd 751b2239e

**Comment by ngoldbaum** [comment]
So there are two issues here. 

One is that in the demeshening we don't have a 3D structure to deposit onto by "default". We used to use the octree. The grid frontends use the AMR mesh. We should probably come up with a way to handle this since it's also a backward compatibility issue.

The other issue is that we don't have a way to generate smoothing lengths for N-body particle data and visualize those data with the SPH pixelizers.

The solution to both problems are probably related, for example we could just deposit onto images, for example, and add the SPH smoothed visualuzation as a new deposit type.

**Comment by sheagk** [comment]
I'm not sure I understand the subtleties of the first issue you mention (specifically, are SPH particles not put onto a 3D structure?), but with regard to the second -- per my discussion on the Slack -- is there a problem with using the distance to the Nth nearest neighbor (where N is an option passed in) to define smoothing length?  You could also allow the user to pass in an array/single value to hard-code the smoothing length too.

**Comment by ngoldbaum** [comment]
There’s no problem with either option. We already have a utility function that does that (calculate smoothing lengths using the distance to the nth nearest neighbor, it’s used in the tipsy frontend because that data format doesn’t have smoothing lengths on dism). We jist need to figure out an appropriate API and wire everything up.

**Comment by qobilidop** [comment]
This might be a bad idea, but I'm wondering if these SPH functionalities could be extracted into a separate package, just like unyt? I feel these functionalites like SPH pixelization and  smoothing length calculation are not tightly coupled to any frontend, or the yt data representation in general. The low-level Cython functions operate on arrays as far as I know. If extracted into a separate package, it would probably be more flexible to use and easier to maintain. This might be a significant amount of work. But as the yt dev workshop is coming, could it be something worth considering?

**Comment by ngoldbaum** [comment]
Perhaps, although I think spelling it "from yt.utilities.sph_utilties import foo" vs "from sph_utilities import foo" doesn't really make much difference. Is there some other advantage you see from factoring this out?

Note also that this package would need to depend on cykdtree at a binary level and would have cython code, so you wouldn't get out of any packaging headaches doing that.

I think it might be a better use of time right now to focus on the API for exposing this functionality. Once we agree on an API we can think about moving code into new packages.

Also having a lower-level "nice" API for SPH smoothing things and generating smoothing lengths doesn't really fix the original issue with ProjectionPlot not knowing how to handle non-SPH particle data, so there would still be work that needs to happen in the high-level yt interface.

**Comment by qobilidop** [comment]
Sorry I deleted my original comment thinking it's immature after posting, which is pasted below:
> This might be a bad idea, but I'm wondering if the SPH-related functionalities, like pixelization, interpolation and smoothing length calculation could be extracted into a separate package, just like unyt? I feel they are not tightly coupled to any frontend, or the yt data representation in general. The low-level Cython functions operate on arrays as far as I know. If extracted into a separate package, it would probably be more flexible to use and easier to maintain. This might be a significant amount of work, and there. But as the yt dev workshop is coming, could it be something worth considering?

Thanks for your kind reply. Yes, I'm mostly thinking about the lower-level API for SPH smoothing on arrays directly. Then a workaround for handling non-SPH particles would be pixelizing with a smoothing length array calculated elsewhere using this lower-level API. It's indeed not solving the high-level interface issue. But having a convinient workarount could be a nice thing.

As I said, I feel my comment is immature because I haven't explored the relevant codebase enough. I'll take a deeper look to have a better idea.

**Comment by ngoldbaum** [comment]
I don't think your comment was immature! I'm sorry my tone made you feel that way, that was not my intent. It's definitely a fair question. I was just trying to explain why I think we need to worry about other things before it makes sense to discuss factoring the low-level SPH utilities into a new package.

**Comment by qobilidop** [comment]
Here's an example of what I'm thinking about: https://gist.github.com/qobilidop/a8c3b4777c94f160442af0446d191ace

For the original issue, I'm proposing using this low-level API to do SPH visualization with arbitrary arrays as a workaround.

**Comment by chummels** [comment]
Fixed by PR #2186
